### ISSUE TITLE
Getting error redux-persist: rehydrate for "root" called after timeout

### ISSUE THREAD

**Issue by ajaymore** [rt2zz/redux-persist#717]
```
redux-persist: rehydrate for "root" called after timeout., Object {
  "INITIALIZING": true,
  "_persist": Object {
    "rehydrated": true,
    "version": -1,
  },
  "user": null,
}, undefined
- node_modules/redux-persist/lib/persistReducer.js:92:59 in _rehydrate
- ... 13 more stack frames from framework internals
```

I am using expo sdk version 25
My code is as below

reducers.js
``` js
export default (
  state = {
    INITIALIZING: true,
    user: null
  },
  action
) => {
  switch (action.type) {
    default:
      return state;
  }
};
```

configureStore.js
``` js
import { createStore } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import rootReducer from './reducers';

const persistConfig = {
  key: 'root',
  storage
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export default () => {
  let store = createStore(persistedReducer);
  let persistor = persistStore(store);
  return { store, persistor };
};

```

App.js

``` js
import React from 'react';
import { StyleSheet, Text, ActivityIndicator, View } from 'react-native';
import * as firebase from 'firebase';
import { PersistGate } from 'redux-persist/integration/react';
import { firebaseConfig } from './src/config';
import { Provider } from 'react-redux';
import configureStore from './src/configureStore';
import styled from 'styled-components';
// import Main from './src/screens/Main';
const { store, persistor } = configureStore();
firebase.initializeApp(firebaseConfig);

const StyledLoaderContainer = styled.View`
  justify-content: center;
  align-items: center;
  margin-top: 100px;
`;

const Loading = () => (
  <StyledLoaderContainer>
    <ActivityIndicator />
    <Text>Initializing...</Text>
  </StyledLoaderContainer>
);

export default class App extends React.Component {
  render() {
    return (
      <Provider store={store}>
        <PersistGate loading={<Loading />} persistor={persistor}>
          <View>
            <Text>This is root</Text>
          </View>
        </PersistGate>
      </Provider>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center'
  }
});
```

**Comment by LeonChabbey** [comment]
Same here.

Seems related to this recent commit: https://github.com/rt2zz/redux-persist/commit/98a78d9aa39d6634a39bd81c776d3ce91e0e7192

**Comment by JanithaR** [comment]
The same happens to me after upgrading to v5.7.1

**Comment by Amurmurmur** [comment]
Same story here :/

**Comment by avid21** [comment]
I just downgraded version to 5.6.12 and still not working

**Comment by NichAga** [comment]
@avid21 Reverting back to 5.6.12 worked for me.

**Comment by avid21** [comment]
@NichAga Yeah it should work. I'm working on it.

**Comment by JanithaR** [comment]
How to downgrade? Uninstall, updated package.json and re-install?

**Comment by krzysztof-miemiec** [comment]
Use semver `~5.6.0` instead of `^5.6.0`. I forgot about it at first.

**Comment by avid21** [comment]
@NichAga Yes, it should work. I was using Remote JS debugging. I stopped it and hence naturally it's working now.

**Comment by NichAga** [comment]
@JanithaR npm install redux-persist@5.6.12
@avid21 good :D

**Comment by diegorodriguesvieira** [comment]
Same here! =(

**Comment by JanithaR** [comment]
Thanks @NichAga I figured it out. Just downgrade and keep working for the moment guys.

**Comment by rohanx96** [comment]
Yes, getting the same error. Reverting to 5.6.12 worked

**Comment by AlexCoderBit** [comment]
Had the same problem, "^5.7.1" FTW!

**Comment by l3lackcurtains** [comment]
Worked after switching back to **redux-persist@5.6.12**

**Comment by outaTiME** [comment]
Same 😡

**Comment by jarvisluong** [comment]
Switch back to 5.6.12 works!

**Comment by outaTiME** [comment]
@jarvisluong yup im switch back to ~5.6.0

**Comment by rt2zz** [comment]
apologize for the churn here, v5.7.2 resolves the issue.

tests missed the issue on 5.7.1 as it was a console.error, not an actual exception

**Comment by RodolfoSilva** [comment]
@rt2zz I still have the same error after update to v5.7.2.

**Comment by sytolk** [comment]
I don't have this error with v.5.7.2. @RodolfoSilva are you sure that you upgrade version? You can check this in npm.lock or yarn.lock file.

**Comment by RodolfoSilva** [comment]
@sytolk, Yes, I'm use the latest version and still have the error. This error only happens when I use the debug in the first time.

```
"redux-persist": {
      "version": "5.7.2",
      "resolved": "https://registry.npmjs.org/redux-persist/-/redux-persist-5.7.2.tgz",
      "integrity": "sha512-NgkSoPA3o6vAo09U5hnzO8CgHNhnBIcZyZBr+tRa/pDDeQqtcsP/zru9b2vz5tIpK6GCYGEsEjSzWUU/vxfRQA=="
    },
```

**Comment by ajaymore** [comment]
I can confirm it's working with 5.7.2.

**Comment by patlux** [comment]
For me it was the short 'debounce' time which caused the error:

```
const persistOptions = {
  storage,
  debounce: 100, // <-- not working in 'Debug JS' mode
  key: DeviceInfo.getBundleId(),
  transforms: [compressor],
};

persistReducer(persistOptions, rootReducer)
```

I increased the time to **500** and now it's working fine.

**Comment by challme28** [comment]
I have to say I've been stuck in this problem for a while now and it was my fault. I forgot to return an action on an Epic. I hope this helps anyone

**Comment by calanti** [comment]
I am stuck with this issue on my recently expo detached project. My state is never persisted regardless of remote debugging, and occasionally I get the error like the OP.

Expo: 24.0.0
react: 16.0.0
RN: Expo 24.0.0 fork
react-redux: 5.0.7
redux: 3.7.2
redux-persist: 5.9.1

On an up note, thank you very much for your work!

**Comment by droegier** [comment]
Same as @calanti

**Comment by kapobajza** [comment]
I was getting the same error on redux-persist version 5.9.1.

It was working fine until I put the `whitelist` prop into `rootPersistConfig `:

`const rootPersistConfig = {
    key: 'root',
    ...
    whitelist: ['auth']
};`

But then I switched back to using the `blacklist` property and everything works fine now.

`const rootPersistConfig = {
    key: 'root',
    ...
    blacklist: ['navigation', 'isLoading', 'hasErrored']
};`

**Comment by calanti** [comment]
Thanks for the insight and yes I am whitelisting, but blacklisting is not really a maintainable solution for me, too many recuders and I only need 2 whitelisted. Hopefully this can help the devs identify the issue though!

**Comment by baba43** [comment]
I have just created a new project with Expo and can not get this work either.

**Comment by ChristienGuy** [comment]
I get this issue with 5.9.1.
For me it only occurs when remote debugging.

Potentially related to this issue with setTimeout in React Native https://github.com/facebook/react-native/issues/4470

**Comment by augusto-altman** [comment]
Same as @ChristienGuy, only get the issue when remote debugging.

**Comment by tarikhagustia** [comment]
same here @ChristienGuy  error when remote debugging :3

**Comment by jamessawyer** [comment]
some issue with redux-persist V5.9.2 when debugger is open.

**Comment by kenkotch** [comment]
After trying a few different versions to no avail, I restarted my computer like I would have on Windows98. Everything is back to normal and working again. It's not a good solution, but may be worth a try if you are still stuck...(I was only getting the error in debug mode)

**Comment by nickofthyme** [comment]
Appears to be working on `5.7.2` with remote JS Debugger.

**Comment by xliank** [comment]
I'm using version `5.9.1` , only show error when enable `Remote JS Debugging`. 
Still no solution?

**Comment by jgfidelis** [comment]
Same issue: Android with Remote debugging on 5.9.1. Will try version 5.7.2.

**Comment by Dmitrylolo** [comment]
same

`"axios": "^0.18.0",
    "expo": "^27.0.2",
    "qs": "^6.5.2",
    "react": "16.3.1",
    "react-native": "https://github.com/expo/react-native/archive/sdk-27.0.0.tar.gz",
    "react-native-elements": "^0.19.1",
    "react-navigation": "^2.0.1",
    "react-redux": "^5.0.7",
    "redux": "^4.0.0",
    "redux-persist": "^5.9.1",
    "redux-thunk": "^2.2.0"`

**Comment by FarhadMahmudi1990** [comment]
using Redux Persist version  5.9.1 ... same issue

**Comment by altenorjr** [comment]
5.10 idem

**Comment by wmlutz** [comment]
"redux-persist": "^5.9.1"

I'm having the same issues. Still no fix? Or hints? Or clues?

**Comment by ilovett** [comment]
@wmlutz restarting the android emulator and/or wiping the data of the device image via AVD Manager

**Comment by MattJakeman** [comment]
Same issue for me on 5.10.0 - It's working correctly until I switch on remote debugging.

**Comment by nilscox** [comment]
I also do encounter this error with `redux-persist@5.10.0`.

The error message I'm getting during the `persist/REHYDRATE` action is `redux-persist: persist timed out for persist key "root"`.

As far as I understand, `redux-persist` internally uses RN's AsyncStorage in order to store the state. Also, there is [an issue](https://github.com/facebook/react-native/issues/12830) when using AsyncStorage when debugging remotly, which makes the promise hang indefinitely (and may cause a timeout).

Could it be the reason why `redux-persist` fails to load the state?

**Comment by Arjith-Natarajan** [comment]
@nilscox are you using any middlewares like redux-logger? I faced the exact same issues as yours while **Debugging JS remotely**

1. `redux-persist` was not working, threw `err : persist timed out for persist key "root"`
2.  AsyncStorage promises never got resolved - [issue](https://github.com/facebook/react-native/issues/12830)

And however when I tried to run with remote debugger it threw an error saying ` TypeError: undefined is not an object (evaluating self.fetch)` -  [issue](https://github.com/facebook/react-native/issues/9599)
even tried to fix it based on this snippet provided [here](https://github.com/davidvuong/ddfa-rn-app/commit/e54a799a92d28a894b7c1ad6d2e3efc727922069)

and once I removed `redux-logger` the app started to behave in expected manner. Promises got resolved without timeout. Somehow made me feel all were connected! :confused: 

Hope it helps someone! :raised_hands: 
Cheers!

**Comment by ilovett** [comment]
I don't necessarily think this is an issue with redux-persist -- I use React Native Debugger, which has an option to `Log AsyncStorage content` to console.

When I first cold boot an emulator and run the project, rehydrate works immediately for a while.  During this time if I `Log AsyncStorage content` then it will output the content to console immediately.

Once rehydrate starts timing out, if I `Log AsyncStorage content` then I get nothing in my console... which makes me think this is more of an android emulator issue or asyncstorage issue...

**Comment by ilovett** [comment]
see: https://github.com/facebook/react-native/pull/18522

**Comment by nilscox** [comment]
@Arjith-Natarajan I am using `redux-logger`. I will investigate if removing it can solve the problem.

@ilovett this is defenetly not a `redux-persist` related error, but it has a side effect that makes `redux-persist` timeout. It happens to me on a real device on Android, so I'm pretty sure it's not related to the emulator environment.

For now, I am using this workaround. I created [this tool](https://github.com/nilscox/rest-storage) that allow to access a storage on my computer from my device. I am using it with a [networkStorage](https://gist.github.com/nilscox/b57be1e47cd785697563c4bbb0917c89) adapter that I give to `redux-persist`.

**Comment by jjercx** [comment]
- `"react-redux": "5.0.7"`
- `"redux": "4.0.0"`
- `"redux-logger": "3.0.6"`
- `"redux-persist": "^5.9.1"`
- `"redux-saga": "0.16.0"`

For me happens debugging and on android device.

**Comment by IOsonoTAN** [comment]
I used `"redux-persist": "^5.10.0"` on Next.js, first i got the same error, i tried to not set initialState or set to object empty `{}` like this:

> const store = createStore(persistedReducer, {})

It persisted and it works find.

**Comment by phuihock** [comment]
I have the same problem. Setting `timeout` > 5s seems to fix it. `timeout` defaults to 5s. I set mine to 10s.

`
const persistConfig = {
    timeout: 10000,
    key: 'root',
    storage,
};
const persistedReducer = persistReducer(persistConfig, reducers);
`

Using redux-persist@5.9.1

**Comment by Arjith-Natarajan** [comment]
:+1 @phuihock that's definitely a fix. I used to get error messages in middle of development, saying `Store rehydrated after time out` , some 20 mins after I had started.

So we still need to figure out what max timeout to set as it varies for everyone.
Still a mystery ...

**Comment by wd** [comment]
As @ChristienGuy said, anyone meet this problem only in remote debugging, should check this comment https://github.com/facebook/react-native/issues/4470#issuecomment-302869193 , try to set the device time is ***exactly*** the same as your computer, or set the timeout more lager like 20 min (caution: set the timeout lager like this may have side effects, only use it when you know what it is )

**Comment by tommyalvarez** [comment]
Happened to me also with the latest version v5.10.0 on web version

**Comment by PaliwalSparsh** [comment]
Same here I am also getting the same error in `"redux-persist": "^5.10.0"`. Please can someone help.

**Comment by tommyalvarez** [comment]
@PaliwalSparsh downgrade to 5.6.12 it will work

**Comment by Nyadesune** [comment]
still happening with 5.10.0, it's a bit weird because it happened after few mins working without issues

**Comment by maxammann** [comment]
Having the same problem on 5.10.0 with react-native and AsyncStorage . Weird that it persists though all these versions.

**Comment by tommyalvarez** [comment]
@ajaymore please reopen the issue

**Comment by maxammann** [comment]
I upgraded to react-native 0.57.0 and removed the `_persist` key from my initial state (I had is there because of flow typing issues).

So probably the removal of the `_persist` key solved it for me.

**Comment by kevupton** [comment]
`Delete the App`
`Reinstall the App`

Works for me.

**Comment by adam-buckley** [comment]
For me I noticed that the console was spamming an endless amount of errors complaining that the my laptop and Android device times had diverged by more than 60 seconds.

Because I dual boot with Linux and Windows, the time on Windows is constantly incorrect. The app was debugging fine and this error seemed to have come out of nowhere. Updating my Windows clock to be the same as my Android device fixed this issue... not really sure why it was a problem in the first place though.

I expect this won't be the solution for anyone at all but still posting just in case.

Redux bits from my package-lock.json, using react-native 0.55.4:
```js
"redux": {
      "version": "4.0.0",
      "resolved": "https://registry.npmjs.org/redux/-/redux-4.0.0.tgz",
      "integrity": "sha512-NnnHF0h0WVE/hXyrB6OlX67LYRuaf/rJcbWvnHHEPCF/Xa/AZpwhs/20WyqzQae5x4SD2F9nPObgBh2rxAgLiA==",
      "requires": {
        "loose-envify": "^1.1.0",
        "symbol-observable": "^1.2.0"
      }
    },
    "redux-axios-middleware": {
      "version": "4.0.0",
      "resolved": "https://registry.npmjs.org/redux-axios-middleware/-/redux-axios-middleware-4.0.0.tgz",
      "integrity": "sha1-gZUcPZrc5vg++JL9FWhXHOvkozY="
    },
    "redux-logger": {
      "version": "3.0.6",
      "resolved": "https://registry.npmjs.org/redux-logger/-/redux-logger-3.0.6.tgz",
      "integrity": "sha1-91VZZvMJjzyIYExEnPC69XeCdL8=",
      "requires": {
        "deep-diff": "^0.3.5"
      }
    },
    "redux-persist": {
      "version": "5.10.0",
      "resolved": "https://registry.npmjs.org/redux-persist/-/redux-persist-5.10.0.tgz",
      "integrity": "sha512-sSJAzNq7zka3qVHKce1hbvqf0Vf5DuTVm7dr4GtsqQVOexnrvbV47RWFiPxQ8fscnyiuWyD2O92DOxPl0tGCRg=="
    }
```

**Comment by ksairi** [comment]
+1 @phuihock 's approach worked for me!

**Comment by MacKentoch** [comment]
@phuihock solution is nice: that make sense to extend timeout value.

**Comment by ksairi** [comment]
Actually, after a day, timeout error appeared again :(

**Comment by ssorallen** [comment]
I increased the timeout in my app to 60s, but in my environment that sort of timeout is still possible. If hitting the timeout actually means "delete the whole state", it'd be nice to have a way to gracefully handle that timeout in my app. Show the PersistGate while retrieving the state and then show a big "Something went wrong, reload the app" if the timeout is hit.

**Comment by cmedinadeveloper** [comment]
Adding timeout time of 10 seconds will work if you are using Debugger

const persistConfig = {
    timeout: 10000,
    key: 'root',
    storage,
};

Tested on redux-persist: 5.10.0

**Comment by ssorallen** [comment]
A workaround I've discovered for my own use: kill redux-persist's `timeout` functionality. If the timeout hits, your store is rehydrated with the initial state and then persisted again to kill the state that is likely still loading.

In your persist config:

```
const persistConfig: {
   ...
   timeout: 0, // The code base checks for falsy, so 0 disables
};
```

A complete solution for redux-persist would be to add a `timeout` component to `PersistGate` that would render if your timeout is it and stop the rehydration work. What it shouldn't do is then rehydrate your store with initialState, which is the current functionality.

**Comment by andrewzey** [comment]
So the best solution for this is in fact to use React Error Boundaries.

If there is an error thrown during the initial render that happens after rehydrate, then the timeout will be reached, and an additional rehydrate will fire that replaces all state with initial state. This results in catastrophic data loss, so doubtful that this is what you want. In a react-native environment, however, NOT doing this means your users will be stuck in a state where the app can't ever start again.

By wrapping the component tree within PersistGate in an error boundary, you can prevent this from occurring by catching and handling the error, and presenting a fallback UI. It's up to you what to do about the defective state tree that results in a crash on the initial render. In my case, I've elected to log the error and provide the user with a choice of submitting feedback and exiting the app (re-throwing an error) and waiting for help, or to clear their data and start from a fresh state. How this should be handled will be highly dependent on the particulars of your application.

I'll provide a code snippet a bit later once I've finalized my own implementation of this, but that should help unblock folks.

**Comment by ssorallen** [comment]
@andrewzey Using an error boundary makes sense, but I can't understand a case where you would actually want catastrophic data loss (i.e. replacing the state with initial state after a timeout is reached). With the worst case you mentioned, "users will be stuck in a state where the app can't ever start again", you can push an update to your app to fix the hydration exception and those users can update the app to get going again without losing any data.

**Comment by andrewzey** [comment]
@ssorallen I tend to agree with you for the vast majority of cases, but there are some cases where having a "Reset state" button would be fine. I often build little test sandboxes where I rely on error boundaries that have "reset" buttons in them. I agree that for most customer-facing scenarios, it doesn't make sense. TBD on whether my suggested "Reset" button will even make it in our app or not.

But the Error Boundary Solution does seem to work well, and it even presents an opportunity to show a feedback form where a user can describe what they were doing when the crash was encountered. That can really help debugging.

That said, I think it's crazy for this to be the default behavior...

**Comment by ssorallen** [comment]
@andrewzey Cool, I do like your approach handling real errors. Thanks for the Error Boundary suggestion.

What about the case where your app hits the timeout but without any exception being thrown. Cases like network connectivity problems, the user goes into a tunnel, turns on airplane mode, or something like that, and your app never throws an exception. Would Error Boundaries be able to solve that?

**Comment by andrewzey** [comment]
@ssorallen touche!

I don't think Error Boundaries could solve for timeouts that get hit when no exception is thrown. So it's probably best to both disable timeout altogether as you suggested (because of it's crazy default behavior that destroys all data) and do an error boundary (if handling render errors separately / outside of any global error handler is desired). 

However, I don't think react apps should wait to render for any kind of network call. That should always happen after initial render in `componentDidMount`, so I'm honestly having a little trouble imagining a scenario where the timeout could get exceeded because of network connectivity problems. Do you have an example in mind?

**Comment by Shanie1331** [comment]
@ssorallen , it works but what will be effected in store ?
please guide

**Comment by ssorallen** [comment]
@Shanie1331 If you set `timeout: 0`, the code that replaces the state with the default state will never run. The exception in the title of this issue will never occur. It won't affect your stored data at all.

If it's possible for your storage to take a long time to respond, redux-persist will simply wait until the storage responds. The `PersistGate` will stay visible until your storage ultimately responds.

**Comment by akiwarheit** [comment]
This may sound stupid but after uninstall, reinstalling, resetting metro, deleting artifacts (typescript), it still doesnt work on Android emulator.

Solution was to wipe the emulator data and start all over again. Emulator is butter smooth after wiping data as well. I think its a timeout/sync issue; the emulator bogs down after running for awhile.

**Comment by rodrigofbm** [comment]
Same here:
store config:

```js
import { applyMiddleware, createStore, compose } from 'redux';
import { offline } from '@redux-offline/redux-offline';
import offlineConfig from '@redux-offline/redux-offline/lib/defaults';
import thunk from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'

import rootReducer from './Reducers';


const persistConfig = {
  key: 'root',
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer)

const store = createStore(
  persistedReducer,
  {},
  compose(
    applyMiddleware(thunk),
    offline(offlineConfig)
  )
);

const persistor = persistStore(store);

export { persistor, store };
```

The problem happens the first time we open the app, if we gave a Reload it works and the error does not come back anymore. But I do not like the first time the user downloads the app it does not run at first.

**Comment by Gustash** [comment]
I was getting the same error. The issue was the emulators date and time weren't synchronised with my computer's. Make sure your date and time is the same between both

**Comment by wkoutre** [comment]
What @Gustash recommends + be sure your time zones are also aligned.

**Comment by sirjmkitavi** [comment]
I just needed to restart my emulator

**Comment by sirjmkitavi** [comment]
I just needed to restart my emulator.
Thanks @gustas

**Comment by sirjmkitavi** [comment]
I just needed to restart my emulator.
Thanks @Gustash

**Comment by dpchami** [comment]
> @NichAga Yes, it should work. I was using Remote JS debugging. I stopped it and hence naturally it's working now.

This one solve my problem

**Comment by hryer** [comment]
try this

`
const persistConfig = {
  key: 'root',
  storage,
  timeout: null
};`

work for me

**Comment by enri90** [comment]
Same problem to react-native 0.60.0

**Comment by NativLouie** [comment]
this solved it for me on version 6.0.0
```

const persistConfig = {
	timeout: 0,
	// Root
	key: 'root',
	// Storage Method (React Native)
	storage: AsyncStorage,
	// Whitelist (Save Specific Reducers)
	stateReconciler: autoMergeLevel2
};
```

**Comment by Zaib01** [comment]
> > @NichAga Yes, it should work. I was using Remote JS debugging. I stopped it and hence naturally it's working now.
> 
> This one solve my problem

**Comment by mbret** [comment]
`persistor.persist()` is still broken since 2018 on react native (did not try on web). 
We have not way of resetting the state (after logout) and purge the store by using the official persistor API. 

The only way we could have a cleanup and a persistence that still works afterward is by doing that hack
https://gist.github.com/mbret/511edf9627b743ad8a52559be61e3d45

Can we make this bug official somehow and work on that ?

**Comment by prateekchachra** [comment]
Still happening with React Native version 0.60.0 and 5.10.0.


I tried the suggestion with timeout being set to 0, not working at all.

**Comment by diosney** [comment]
I updated to redux-persist 6.0.0 and it seems fixed.

**Comment by AgustinJimenez** [comment]
> try this
> 
> ` const persistConfig = { key: 'root', storage, timeout: null };`
> 
> work for me

+1

**Comment by 256hz** [comment]
I got this error when the device I was testing on didn't have internet connectivity, and went away when it did.  Could be related to the time mismatch issue @Gustash noted above.

**Comment by PedroBern** [comment]
Got this error on `v6` during remote debugging because my computer time was 1 hour behind. After fixing the time, the error was gone!

**Comment by hao1234567** [comment]
> Got this error on `v6` while remote debugging because my computer time was 1 hour behind. After fix the time, the error was gone!

Thank you for reminding me!This worked for me!

**Comment by sintylapse** [comment]
For me issue appeared after upgrading RN to 0.60.0 and react-native-navigation to v4 and was resolved after I replaced previously recommended by wix android run script
```
"android": "cd ./android && ./gradlew app:assembleDebug && ./gradlew installDebug",
```
with the standard script
```
"android": "react-native run-android",
```

**Comment by vsla** [comment]
> Adding timeout time of 10 seconds will work if you are using Debugger
> 
> const persistConfig = {
> timeout: 10000,
> key: 'root',
> storage,
> };
> 
> Tested on redux-persist: 5.10.0

Worked for me, tanks!

**Comment by renielDev** [comment]
> A workaround I've discovered for my own use: kill redux-persist's `timeout` functionality. If the timeout hits, your store is rehydrated with the initial state and then persisted again to kill the state that is likely still loading.
> 
> In your persist config, prevent the [timeout setup (persistReducer.js#L88-99)](https://github.com/rt2zz/redux-persist/blob/v5.10.0/src/persistReducer.js#L88-L99):
> 
> ```
> const persistConfig: {
>    ...
>    timeout: 0, // The code base checks for falsy, so 0 disables
> };
> ```
> 
> A complete solution for redux-persist would be to add a `timeout` component to `PersistGate` that would render if your timeout is hit and stop the rehydration work. What it shouldn't do is then rehydrate your store with initialState, which is the current functionality.

What will be the implications of doing this?
I've been having similar issue and `setting timeout: 0` fixed it. But I'm worried that there's a catch?

**Comment by ravindranpandu** [comment]
This issue was been there for a while, i tried a lot fixing it and finally @Gustash solution resolves it, thank you very much 👍

**Comment by rueiwoqp** [comment]
FYI, setting timeout to null instead of 0 will not fix this issue.  Timeout must be a number. You can run the tests and verify if you want.  Here's what the test output says when timeout is null:

```➜  redux-persist git:(master) ✗ npm run test

> redux-persist@6.0.0 test
> flow && cross-env BABEL_ENV=commonjs ava

Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ tests/complete.spec.js:42:74

Cannot call persistReducer with object literal bound to config because null [1] is incompatible with number [2] in
property timeout.

     tests/complete.spec.js
     39│
     40│ test('persistStore timeout 0 never bootstraps', t => {
     41│   return new Promise((resolve, reject) => {
 [1] 42│     let r1 = persistReducer({...config, storage: brokenStorage, timeout: null}, reducer)
     43│     const rootReducer = combineReducers({ r1 })
     44│     const store = createStore(rootReducer)
     45│     const persistor = persistStore(store, null, () => {

     src/types.js
 [2] 28│   timeout?: number,```

**Comment by SudoPlz** [comment]
Just to make this clear, is this a dev only issue (meaning that if we don't use any timeout values users won't have issues in production will they?)

**Comment by rueiwoqp** [comment]
> Just to make this clear, is this a dev only issue (meaning that if we don't use any timeout values users won't have issues in production will they?)

This issue exists in all environments.

**Comment by andrewzey** [comment]
@SudoPlz Definitely an issue affecting production. I recommend thoroughly combing through the redux-persist code. Complete and total data erasure on production clients will be very frustrating, and it's a very good idea to understand the data storage layer fully.

If we weren't planning on moving away from react-native to native, we would honestly re-write our own lib to handle this because redux-persist has a few issues and it's not actively maintained anymore.

**Comment by iagormoraes** [comment]
> @SudoPlz Definitely an issue affecting production. I recommend thoroughly combing through the redux-persist code. Complete and total data erasure on production clients will be very frustrating, and it's a very good idea to understand the data storage layer fully.
> 
> If we weren't planning on moving away from react-native to native, we would honestly re-write our own lib to handle this because redux-persist has a few issues and it's not actively maintained anymore.

as the package is not maintained anymore we could at least have some examples on how to migrate the data to another source, it will affect tons of projects.

**Comment by andrewzey** [comment]
> > @SudoPlz Definitely an issue affecting production. I recommend thoroughly combing through the redux-persist code. Complete and total data erasure on production clients will be very frustrating, and it's a very good idea to understand the data storage layer fully.
> > If we weren't planning on moving away from react-native to native, we would honestly re-write our own lib to handle this because redux-persist has a few issues and it's not actively maintained anymore.
> 
> as the package is not maintained anymore we could at least have some examples on how to migrate the data to another source, it will affect tons of projects.

As best I can tell that would require writing an alternative to redux-persist, understanding its persistence interaction with the file system / async storage, then writing a migration guide. That seems an unreasonable expectation of an author who abandoned the old project. Much more likely that a new project taking over the problem of redux-persistence would write a guide / utility for migrating TO their new project.

**Comment by tanveer-ahmad74** [comment]
> @Avid21 Reverting back to 5.6.12 worked for me.

not working for me and 6.0.0 as well

**Comment by nickensoul** [comment]
Still actual

**Comment by FreddyCarrillo** [comment]
Lo que funcionó para mi fue pasar directamente en el reducer el persistedReducer

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
	reducer: persistedReducer,

**Comment by 002love** [comment]
Thanks, @hryer 
Also in current versions of redax-perist, you will have to specify the time as a number in milliseconds, "20000" works fine on most devices
